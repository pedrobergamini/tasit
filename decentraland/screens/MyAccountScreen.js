import React from "react";
import MyAccount from "@presentational/MyAccount";
import { connect } from "react-redux";
import AccountCreationActions from "@constants/AccountCreationActions";
import ActionStatus from "../constants/ActionStatus";
import PropTypes from "prop-types";

export class MyAccountScreen extends React.Component {
  render() {
    const { accountInfo } = this.props;
    const creationActions = [];
    creationActions.push({
      name: "Account created",
      action: null,
      status: !accountInfo.account ? ActionStatus.MISSING : ActionStatus.DONE,
    });
    Object.keys(AccountCreationActions).forEach(action => {
      const creationAction = {
        name: AccountCreationActions[action].name,
        action,
      };
      // TODO: As soon as we store action status in redux, this logic will change
      // transaction pending, confirmed once, confirmed many times, failed, etc.
      if (accountInfo.creationActions.hasOwnProperty(action)) {
        creationAction.status = ActionStatus.DONE;
      } else {
        creationAction.status = ActionStatus.MISSING;
      }
      creationActions.push(creationAction);
    });
    const isAccountCreated = !!accountInfo.account;
    return (
      <MyAccount
        progress={this._getPercentage(isAccountCreated, creationActions)}
        creationActions={creationActions}
      />
    );
  }

  _getPercentage(isAccountCreated, accountActions) {
    let percentage = isAccountCreated ? 0.25 : 0;
    accountActions.forEach(accountAction => {
      const isDone = accountAction.status === ActionStatus.DONE;
      const hasDefinedActionPercentage = AccountCreationActions.hasOwnProperty(
        accountAction.action
      );
      if (isDone && hasDefinedActionPercentage) {
        percentage += AccountCreationActions[accountAction.action].percentage;
      }
    });
    return percentage;
  }
}

MyAccountScreen.propTypes = {
  accountInfo: PropTypes.object,
};

const mapStateToProps = state => {
  const { accountInfo } = state;
  return { accountInfo };
};

export default connect(mapStateToProps)(MyAccountScreen);
