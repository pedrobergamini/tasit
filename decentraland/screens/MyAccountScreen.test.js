import React from "react";
import { shallow } from "enzyme";
import { MyAccountScreen } from "./MyAccountScreen";
import AccountCreationStatus from "@constants/AccountCreationStatus";

describe("MyAccountScreen", () => {
  describe("renders the component", () => {
    it("without account", async () => {
      expect(
        shallow(
          <MyAccountScreen
            accountInfo={{
              account: null,
              creationActions: {},
            }}
          />
        )
      ).toMatchSnapshot();
    });

    it("with account", async () => {
      expect(
        shallow(
          <MyAccountScreen
            accountInfo={{
              account: {},
              creationActions: {
                [AccountCreationStatus.FUNDING_WITH_ETH]: {},
              },
            }}
          />
        )
      ).toMatchSnapshot();
    });
  });
});
