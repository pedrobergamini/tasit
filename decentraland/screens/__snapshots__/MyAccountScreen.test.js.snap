// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAccountScreen renders the component with account 1`] = `
<MyAccount
  creationActions={
    Array [
      Object {
        "action": null,
        "name": "Account created",
        "status": "done",
      },
      Object {
        "action": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "status": "done",
      },
      Object {
        "action": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "status": "missing",
      },
      Object {
        "action": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "status": "missing",
      },
    ]
  }
  progress={0.5}
/>
`;

exports[`MyAccountScreen renders the component without account 1`] = `
<MyAccount
  creationActions={
    Array [
      Object {
        "action": null,
        "name": "Account created",
        "status": "missing",
      },
      Object {
        "action": "FUNDING_WITH_ETH",
        "name": "Funded with ETH",
        "status": "missing",
      },
      Object {
        "action": "FUNDING_WITH_MANA",
        "name": "Funded with MANA tokens",
        "status": "missing",
      },
      Object {
        "action": "APPROVING_MARKETPLACE",
        "name": "Linked to marketplace",
        "status": "missing",
      },
    ]
  }
  progress={0}
/>
`;
